torch_np/_helpers.py

```diff
@@ -29,18 +29,33 @@ def cast_and_broadcast(tensors, out, casting):
     if out is None:
         return tensors
     else:
-        from ._ndarray import asarray, ndarray
-
-        if not isinstance(out, ndarray):
-            raise TypeError("Return arrays must be of ArrayType")
-
         tensors = _util.cast_and_broadcast(
             tensors, out.dtype.type.torch_dtype, out.shape, casting
         )
 
     return tuple(tensors)
 
 
+def ufunc_preprocess(
+    tensors, out, where, casting, order, dtype, subok, signature, extobj
+):
+    # internal preprocessing or args in ufuncs (cf _unary_ufuncs, _binary_ufuncs)
+    if order != "K" or not where or signature or extobj:
+        raise NotImplementedError
+
+    # XXX: dtype=... parameter
+    if dtype is not None:
+        raise NotImplementedError
+
+    out_shape_dtype = None
+    if out is not None:
+        out_shape_dtype = (out.get().dtype, out.get().shape)
+
+    tensors = _util.cast_and_broadcast(tensors, out_shape_dtype, casting)
+
+    return tensors
```

Comment:
Not quite. In numpy, `out` array broadcasts ufunc inputs, and is never resized:

```
In [1]: import torch

In [2]: t = torch.ones(2, dtype=int)

In [3]: out = torch.empty((2, 2))

In [4]: res = torch.add(t, t, out=out)
<ipython-input-4-1158967e8a80>:1: UserWarning: An output with one or more elements was resized since it had shape [2, 2], which does not match the required output shape [2]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at /home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670076225403/work/aten/src/ATen/native/Resize.cpp:17.)
  res = torch.add(t, t, out=out)

In [5]: res
Out[5]: tensor([2., 2.])

In [6]: import numpy as np

In [7]: np.add(t.numpy(), t.numpy(), out=np.empty((2, 2)))
Out[7]: 
array([[2., 2.],
       [2., 2.]])
```

Reply:
ugh, that's incredibly annoying.

Reply:
Actually, can't we simply do as we do in PyTorch but first broadcasting the output of the function to the size of the `out=` tensor before performing the `copy_`? That would allow us to implement the `out=` semantics separately from all the rest, which should be much simpler. This uses the (trivial) fact that broadcasting commutes with pointwise ops.

Reply:
We'll still have to manually do type casting in the preprocessing though. Specifically, this loop, https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/main/torch_np/_detail/_util.py#L198 cannot be avoided, no matter where precisely in lands.

Whether this is going to be much simpler than https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/89 or not, I frankly don't know.  Where is the expected simplification I'm not seeing?

Reply:
How does it show it, mind elaborating?

Reply:
Ah, thanks. I think it's the same thing as these tests in gh-89: https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/89/files

I believe (cf comments in tests) what we have now is a superset of what numpy is doing, esp for binary ufuncs.

Reply:
Also, here's one more snippet of why we need at least some pre-processing:

```
In [31]: np.sin(True, out=np.empty(1, dtype=bool))
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
Input In [31], in <cell line: 1>()
----> 1 np.sin(True, out=np.empty(1, dtype=bool))

UFuncTypeError: Cannot cast ufunc 'sin' output from dtype('float32') to dtype('bool') with casting rule 'same_kind'

In [32]: np.sin(True, out=np.empty(1, dtype=bool), casting='unsafe')
Out[32]: array([ True])
```

Reply:
I don't think that that example proves we need preprocessing. The semantics for those two examples are:
- Run he function
- Check whether the output dtype can be cast to the dtype of `out`
- If not, check whether `casting="unsafe"` otherwise raise an error
And all this is done after running the function.

Reply:
First, there are more casting modes:

```
In [39]: np.sin(True, out=np.empty(1, dtype=bool), casting='no')
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
Input In [39], in <cell line: 1>()
----> 1 np.sin(True, out=np.empty(1, dtype=bool), casting='no')

UFuncTypeError: Cannot cast ufunc 'sin' input from dtype('bool') to dtype('float32') with casting rule 'no'
```

Second, what we are discussing now is I believe a (bunch of) corner case(s). What's more interesting is binary ufuncs. For one,

```
In [41]: np.add(True, 2, casting='no')
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
Input In [41], in <cell line: 1>()
----> 1 np.add(True, 2, casting='no')

UFuncTypeError: Cannot cast ufunc 'add' input 0 from dtype('bool') to dtype('int64') with casting rule 'no'
```

Reply:
All those other casting modes should also be performed on the result, not on the out args. The casting args may affect teh casting of the inputs, but they affect them irregardless of whether the `out=` kwarg is present or not.
Let's have a vc to discuss this on Wed, I think it'll be easier.